src/main/scala/com.snowplowanalytics.sauna/responders/urbanairship/UAResponder.scala

```diff
@@ -0,0 +1,103 @@
+package com.snowplowanalytics.sauna
+package responders
+package urbanairship
+
+import scala.io.Source.fromInputStream
+
+import java.io.InputStream
+
+import apis.urbanAirship
+
+import akka.actor.{ ActorRef, Props }
+
+import responders.Responder.FileAppeared
+
+  /**
+ * Does stuff for UrbanAirship Static List upload feature.
+ *
+ * @see https://github.com/snowplow/sauna/wiki/Urban-Airship-Responder-user-guide
+ * @param urbanairship Instance of UrbanAirship.
+ * @param logger A logger actor.
+ */
+
+class UAResponder(urbanairship: urbanAirship)(implicit logger: ActorRef) extends Responder {
+
+  import UAResponder._
+
+  val pathPattern =
+    """.*com\.urbanairship/
+      |static_lists/
+      |v1/
+      |tsv:\*/
+      |.+$
+    """.stripMargin
+      .replaceAll("[\n ]", "")
+
+  def process(fileAppeared: FileAppeared): Unit = {
+    import fileAppeared._
+    convertTSV(is)
+  }
+  
+/**
+   * Converts the TSV data and groups it in a map so as to make batched upload requests to UrbanAirship
+   *
+   * @param is the InputStream of the TSV file 
+   * @return a map of the data grouped by application keys and then by listNames
+   */
+
+  def convertTSV(is: InputStream): Unit = {
+
+    var appToListMap = Map[String, Map[String, List[Airship]]]()
+
+    var listToIdentifierMap = Map[String, List[Airship]]()
+
+    var count = 0
```

Comment:
Second, it's still pretty big and consist of many mutable variables. 
Mutable variables not a big problem here, because they're not shared between threads, but we can make it more concise and readable using immutable vals.
My (very) personal rule of thumb - 3-8 LOCs for function as nicest length. This is of course by no means requirement for you, but we'll try to stick with it and you'll see great improvement.
Don't forget, everything you do with mutable vars and collections can be done using immutable ones. In 98% of cases it will be more readable and robust solution. Try to avoid mutability if you don't have specific reasons.